cores/esp32/esp32-hal-ledc.c

```diff
@@ -157,110 +216,80 @@ uint32_t ledcWriteNote(uint8_t chan, note_t note, uint8_t octave){
-}
+    ledc_channel_handle_t bus = (ledc_channel_handle_t)perimanGetPinBus(pin, ESP32_BUS_TYPE_LEDC);
+    if(bus != NULL){
 
-uint32_t ledcChangeFrequency(uint8_t chan, uint32_t freq, uint8_t bit_num)
-{
-    if(chan >= LEDC_CHANNELS || bit_num > LEDC_MAX_BIT_WIDTH){
-        log_e("LEDC channel not available! (maximum %u) or bit width too big (maximum %u)", LEDC_CHANNELS, LEDC_MAX_BIT_WIDTH);
-        return 0;
-    }
-    uint8_t group=(chan/8), timer=((chan/2)%4);
-
-    ledc_timer_config_t ledc_timer = {
-        .speed_mode       = group,
-        .timer_num        = timer,
-        .duty_resolution  = bit_num,
-        .freq_hz          = freq, 
-        .clk_cfg          = LEDC_DEFAULT_CLK
-    };
-
-    if(ledc_timer_config(&ledc_timer) != ESP_OK)
-    {
-        log_e("ledcChangeFrequency failed!");
-        return 0;
+        if(resolution > LEDC_MAX_BIT_WIDTH){
+            log_e("LEDC resolution too big (maximum %u)", LEDC_MAX_BIT_WIDTH);
+            return 0;
+        }
+        uint8_t group=(bus->channel/8), timer=((bus->channel/2)%4);
+
+        ledc_timer_config_t ledc_timer = {
+            .speed_mode       = group,
+            .timer_num        = timer,
+            .duty_resolution  = resolution,
+            .freq_hz          = freq, 
+            .clk_cfg          = LEDC_DEFAULT_CLK
+        };
+
+        if(ledc_timer_config(&ledc_timer) != ESP_OK)
+        {
+            log_e("ledcChangeFrequency failed!");
+            return 0;
+        }
+        bus->channel_resolution = resolution;
+        return ledc_get_freq(group,timer);
     }
-    channels_resolution[chan] = bit_num;
-    return ledc_get_freq(group,timer);
+    return 0;
 }
 
-static int8_t pin_to_channel[SOC_GPIO_PIN_COUNT] = { 0 };
-static int cnt_channel = LEDC_CHANNELS;
 static uint8_t analog_resolution = 8;
 static int analog_frequency = 1000;
 void analogWrite(uint8_t pin, int value) {
   // Use ledc hardware for internal pins
   if (pin < SOC_GPIO_PIN_COUNT) {
-    if (pin_to_channel[pin] == 0) {
-      if (!cnt_channel) {
-          log_e("No more analogWrite channels available! You can have maximum %u", LEDC_CHANNELS);
-          return;
-      }
-      if(ledcSetup(cnt_channel - 1, analog_frequency, analog_resolution) == 0){
-          log_e("analogWrite setup failed (freq = %u, resolution = %u). Try setting different resolution or frequency");
-          return;
-      }
-      ledcAttachPin(pin, cnt_channel - 1);
-      pin_to_channel[pin] = cnt_channel--;
+    ledc_channel_handle_t bus = (ledc_channel_handle_t)perimanGetPinBus(pin, ESP32_BUS_TYPE_LEDC);
+    if(bus == NULL && perimanSetPinBus(pin, ESP32_BUS_TYPE_INIT, NULL)){
+
+        if(ledcAttach(pin, analog_frequency, analog_resolution) == 0){
+            log_e("analogWrite setup failed (freq = %u, resolution = %u). Try setting different resolution or frequency");
+            return;
+        }
     }
-    ledcWrite(pin_to_channel[pin] - 1, value);
+    ledcWrite(pin, value);
```

Comment:
One issue I think that there is with analogWrite() would be that it allocates ("eats") ledc channels and there is no way to release them.
A suggestion is to allow it to release/detach the channel when <value> is Zero.
It may have to make sure that the GPIO goes to low level...

Reply:
> I think that analogWrite() has a potential issue:
This function allocates ("eats") LEDC channels and there is no way to release them.

You may be wrong with this, as the peripheral manager will be implemented in 3.0.0, if you setup any other peripheral on that pin, the peripheral manager will call ledcDetachBus and free that LEDC channel.

Reply:
I agree with @P-R-O-C-H-Y here. It's not necessary to switch to GPIO if the result is what is expected to be. Maybe there is a hardware reason to be done in that way on AVR